network: make Address object more consistent with assigned address

yuwata opened the pull request:
Split-out of #28325.

Fixes #25843.

bluca reviewed (changes requested):
please add a commit message explaining why this is done, it's very unclear at the moment

yuwata commented:
> please add a commit message explaining why this is done, it's very unclear at the moment

Indeed. Now I added more corresponding commits, and updated the commit message. PTAL.

bluca reviewed (approved):
typos in the commit messages: "much" -> "more"
